src/IdentityServer/Services/IdsrvResourceStore.cs

```diff
@@ -31,30 +31,34 @@ private Identity.Clients.Models.Resource[] GetAll()
             return _all;
         }
 
-        public async Task<ApiResource> FindApiResourceAsync(string name)
+         public async Task<IEnumerable<ApiResource>> FindApiResourcesByNameAsync(IEnumerable<string> apiResourceNames)
         {
             await Task.Delay(0);
             return GetAll()
-                .Where(r => r.Type == ResourceType.Api && r.Name == name)
-                .Select(r => ConvertApiResource(r))
-                .FirstOrDefault();
+                .Where(r => r.Type == ResourceType.Api && apiResourceNames.Contains(r.Name))
+                .Select(r => ConvertApiResource(r));
         }
 
-        public async Task<IEnumerable<ApiResource>> FindApiResourcesByScopeAsync(IEnumerable<string> scopeNames)
+        public async Task<IEnumerable<ApiResource>> FindApiResourcesByScopeNameAsync(IEnumerable<string> scopeNames)
         {
             await Task.Delay(0);
-            List<ApiResource> result = new List<ApiResource>();
-            var all = GetAll();
-            var map = all.Where(r => r.Type == ResourceType.Api).SelectMany(r => r.Claims).ToDictionary(c => c.Type);
-            foreach(string scope in scopeNames)
-            {
-                if (map.ContainsKey(scope))
-                    result.Add(ConvertApiResource(all.Where(r => r.Id == map[scope].ResourceId).First()));
-            }
-            return result;
+            return GetAll()
+                .Where(r => r.Type == ResourceType.Api && r.Scopes.Split(' ').Any(s => scopeNames.Contains(s)))
+                .Select(r => ConvertApiResource(r));
+        }
+
+        async Task<IEnumerable<ApiScope>> IResourceStore.FindApiScopesByNameAsync(IEnumerable<string> scopeNames)
+        {
+            await Task.Delay(0);
+            return GetAll()
+                .Where(r => r.Type == ResourceType.Api)
```

Comment:
Using `.Distinct()` here because ID4 will complain/fail if there are duplicates. The prefix method (`<resource_name>-<scope_name>`) we are using almost prevents duplicate scopes, but there are cases where there can be duplicates.
For example, two resources: `dev` and `dev-api`. 
`dev` could have scopes `dev, dev-api` (as `dev-__` is a valid scope name here)
`dev-api` could have scopes `dev-api `
Not sure if this is an issue, but thought I'd mention the possibility.

Reply:
We could add a validation to ensure a scope with the same name doesn't exist or change something about how we prefix or ignore/allow this to happen.